src/base.jl

```diff
@@ -280,15 +284,116 @@ function add_component!(sys::ConcreteSystem, component::T) where T <: Component
+    match = true
+
+    for (issue_time, forecasts) in system_forecasts
+        unique_components = Set{Tuple{<:Component, String}}()
+        num_not_unique = 0
+        for forecast in forecasts
+            component_label = (forecast.component, forecast.label)
+            if component_label in unique_components
+                num_not_unique += 1
+                match = false
+                @error("not all components in forecast vector are unique", component_label,
+                       issue_time)
+            else
+                push!(unique_components, component_label)
+            end
+        end
+        @info "summary" num_not_unique length(forecasts)
+    end
+
+    return match
+end
+
+function add_forecast!(forecasts::SystemForecasts, issue_time::IssueTime,
+                       forecast::Forecast)
+    if !haskey(forecasts, issue_time)
+        forecasts[issue_time] = Forecasts{Forecast}()
+    end
+
+    push!(forecasts[issue_time], forecast)
+end
+
+function add_forecast!(forecasts::SystemForecasts, forecast::Forecast)
+    issue_time = get_issue_time(forecast)
+    add_forecast!(forecasts, issue_time, forecast)
+end
+
 """
 Args:
     A ConcreteSystem struct
     A :Symbol=>Array{ <: Forecast,1} Pair denoting the forecast name and array of device forecasts
 Returns:
     A System struct with a modified forecasts field
 """
-function add_forecast!(sys::ConcreteSystem,fc::Pair{Symbol,Array{Forecast,1}})
-    sys.forecasts[fc.first] = fc.second
+function add_forecast!(sys::ConcreteSystem, fc::Pair{Symbol, Vector{Forecast}})
+    tmp = SystemForecastsDeprecated()
+    tmp[fc.first] = fc.second
+    system_forecasts = SystemForecasts()
+    _convert_forecasts!(system_forecasts, tmp)
+
+    for (issue_time, forecasts) in system_forecasts
+        # This assumes that issue_time is the same for each forecast in the vector because
+        # that is how _convert_forecasts works above.
+        for forecast in forecasts
+            add_forecast!(sys.forecasts, issue_time, forecast)
+        end
+    end
+
+    if !_validate_component_label_uniqueness(sys.forecasts)
+        throw(DataFormatError("components and labels are not unique within forecast array"))
+    end
+end
+
+"""Returns an iterator to the forecast IssueTime values stored in the System."""
+function get_forecast_issue_times(sys::ConcreteSystem)
+    return keys(sys.forecasts)
+end
+
+"""Returns an iterator to the forecasts for the given IssueTime stored in the System."""
+function get_forecasts(sys::ConcreteSystem, issue_time::IssueTime)
+    if !haskey(sys.forecasts, issue_time)
+        error("forecast issue_time {issue_time} does not exist")
+    end
+
+    return Iterators.take(sys.forecasts[issue_time], length(sys.forecasts[issue_time]))
+end
+
+"""Returns a vector of forecasts that match the components and label."""
+function get_forecasts(sys::ConcreteSystem, issue_time::IssueTime, components_iterator,
```

Comment:
This implementation might not be exactly what we want.  It will return all the forecasts that match the components and label.  However, it will not throw an error if it does not find a forecast matching one of the components. Should it?

Reply:
In my opinion, yes, it should throw an error if it doesn’t find a forecast for one of the items in the iterator

Reply:
What about the scenario where the user passes the result of get_components() to get_forecasts()?  They could pass something very broad like Component to get_components and get back components that are not tied to forecasts, right?  In that case they may not care if some are not matched.  I have no problem restricting the behavior, but just want to ensure all use cases are covered.

Reply:
That seems valid. Perhaps the default should error if some components aren’t matched with forecasts and we could have a kwarg like filter=false to enable the behavior you mention

Reply:
That seems valid. Perhaps the default should error if some components aren’t matched with forecasts and we could have a kwarg like filter=false to enable the behavior you mention